src/sys/macos.rs

```diff
@@ -0,0 +1,133 @@
+use std::{error, fmt};
+
+use core_graphics::{
+    display::{CGDisplayMoveCursorToPoint, CGError, CGMainDisplayID},
+    event::{CGEvent, CGEventTapLocation, CGEventType, CGMouseButton, ScrollEventUnit},
+    event_source::{CGEventSource, CGEventSourceStateID},
+    geometry::CGPoint,
+};
+
+use crate::types::{keys::*, Point};
+
+impl From<CGPoint> for Point {
+    fn from(other: CGPoint) -> Point {
+        Point {
+            x: other.x as _,
+            y: other.y as _,
+        }
+    }
+}
+
+impl Into<CGPoint> for Point {
+    fn into(self) -> CGPoint {
+        CGPoint::new(self.x as _, self.y as _)
+    }
+}
+
+#[derive(Debug)]
+pub enum Error<'a> {
+    CGDisplayMoveCursorToPoint(CGError),
+    CGEventNotCreated,
+    CGEventSourceStateInvalid,
+
+    InvalidButtonStr(&'a str),
+}
+
+impl<'a> error::Error for Error<'a> {}
+
+impl<'a> fmt::Display for Error<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Error::CGDisplayMoveCursorToPoint(err) => {
+                write!(f, "error in call to CGDisplayMoveCursorToPoint: {}", err)
+            }
+            Error::CGEventNotCreated => write!(f, "CGEvent could not be created"),
+            Error::CGEventSourceStateInvalid => write!(f, "invalid CGEventSourceStateID"),
+
+            Error::InvalidButtonStr(button) => write!(f, "invalid button str: {:?}", button),
+        }
+    }
+}
+
+pub struct Mouse;
+
+impl Mouse {
+    fn event_source<'a>() -> Result<CGEventSource, Error<'a>> {
+        Ok(
+            CGEventSource::new(CGEventSourceStateID::CombinedSessionState)
+                .or(Err(Error::CGEventSourceStateInvalid))?,
+        )
+    }
+
+    pub fn new() -> Mouse {
+        Mouse
+    }
+
+    pub fn move_to(&self, x: i32, y: i32) -> Result<(), Box<dyn error::Error>> {
+        let point = CGPoint::new(x as _, y as _);
+
+        match unsafe { CGDisplayMoveCursorToPoint(CGMainDisplayID(), point) } {
+            0 => Ok(()),
+            err => Err(Box::new(Error::CGDisplayMoveCursorToPoint(err))),
+        }
+    }
+
+    pub fn press<'a>(&self, button: &'a str) -> Result<(), Box<dyn error::Error + 'a>> {
+        let (event_type, mouse_button) = match button {
+            _ if button == LEFT => Ok((CGEventType::LeftMouseDown, CGMouseButton::Left)),
+            _ if button == MIDDLE => Ok((CGEventType::LeftMouseDown, CGMouseButton::Center)),
+            _ if button == RIGHT => Ok((CGEventType::RightMouseDown, CGMouseButton::Right)),
+            _ => Err(Box::new(Error::InvalidButtonStr(button))),
```

Comment:
I'm not that experienced in rust yet but may I ask why this is better than doing?

```rust

        let (event_type, mouse_button) = match button {
            LEFT => Ok((CGEventType::LeftMouseDown, CGMouseButton::Left)),
            MIDDLE => Ok((CGEventType::LeftMouseDown, CGMouseButton::Center)),
            RIGHT => Ok((CGEventType::RightMouseDown, CGMouseButton::Right)),
            _ => Err(Box::new(Error::InvalidButtonStr(button))),
        }

Reply:
Heh, the problem with your snippet is that `LEFT`, despite it being a predefined `const`, doesn't restrict the pattern like in OCaml. Instead the string slice being matched  is bound to a new variable called `LEFT`, regardless of its actual contents. In other words, the latter three patterns are unreachable while `"left"`, `"right"`, and even `"🍆"` matches to a left button press.

It would be possible to do something like this instead,

```
let (event_type, mouse_button) = match button {
    "left" => Ok((CGEventType::LeftMouseDown, CGMouseButton::Left)),
    "middle" => Ok((CGEventType::LeftMouseDown, CGMouseButton::Center)),
    "right" => Ok((CGEventType::RightMouseDown, CGMouseButton::Right)),
    _ => Err(Box::new(Error::InvalidButtonStr(button))),
}
```

but that kind of defeats the purpose of declaring the `const`s in the first place.

I think it would be better to declare an enum instead, so something like,
```
enum Keys {
    Left,
    Right,
    // ...
}
```
and match over that, instead of taking a `&str`. That would also mean no panics, and in my opinion cleaner code. 😉

But I'll leave that up to you. I thought about including it in this PR or creating another, but it felt like a bigger overhaul than needed.

Reply:
Thank you for all the info 😄